File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/executing/DescendentsEvaluator.java
package org.opencds.cqf.cql.engine.elm.executing;

import org.opencds.cqf.cql.engine.runtime.Interval;
import org.opencds.cqf.cql.engine.runtime.Tuple;

import java.util.ArrayList;
import java.util.List;

public class DescendentsEvaluator {

    private static List<Object> descendents = new ArrayList<>();

    public static Object descendents(Object source) {
        if (source == null) {
            return null;
        }

        return getDescendents(source);
    }

    public static Object getDescendents(Object source) {
        if (source instanceof Iterable) {
            for (Object element : (Iterable<?>) source) {
                descendents.add(getDescendents(element));
            }
        }

        else if (source instanceof Tuple) {
            for (Object element : ((Tuple) source).getElements().values()) {
                descendents.add(getDescendents(element));
            }
        }

        else if (source instanceof Interval) {
            descendents.add(getDescendents(((Interval) source).getStart()));
            descendents.add(getDescendents(((Interval) source).getEnd()));
        }

        else {
            descendents.add(source);
        }

        return descendents;
    }

}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/DescendentsEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.opencds.cqf.cql.engine.runtime.Interval
import org.opencds.cqf.cql.engine.runtime.List
import org.opencds.cqf.cql.engine.runtime.Tuple
import org.opencds.cqf.cql.engine.runtime.Value
import org.opencds.cqf.cql.engine.runtime.toCqlList

object DescendentsEvaluator {

    @JvmStatic
    fun descendents(source: Value?): List? {
        if (source == null) {
            return null
        }

        return getDescendents(source)
    }

    fun getDescendents(source: Value?): List {
        val descendents = mutableListOf<Value?>()

        if (source is List) {
            for (element in source) {
                descendents.add(getDescendents(element))
            }
        } else if (source is Tuple) {
            for (element in source.elements.values) {
                descendents.add(getDescendents(element))
            }
        } else if (source is Interval) {
            descendents.add(getDescendents(source.start))
            descendents.add(getDescendents(source.end))
        } else {
            descendents.add(source)
        }

        return descendents.toCqlList()
    }
}
